### ISSUE TITLE
Pods with PVCs stuck in ContainerCreating after Node upgrade

### ISSUE THREAD

**Issue by enisoc** [kubernetes/kubernetes#37585]
**Is this a request for help?** (If yes, you should use our troubleshooting guide and community support channels, see http://kubernetes.io/docs/troubleshooting/.):

No.

**What keywords did you search in Kubernetes issues before filing this one?** (If you have found any duplicates, you should instead reply there.):

* upgrade volume
* mount timeout
* ContainerCreating

---

**Is this a BUG REPORT or FEATURE REQUEST?** (choose one):

BUG REPORT

**Kubernetes version** (use `kubectl version`):

```
Server Version: version.Info{Major:"1", Minor:"5+", GitVersion:"v1.5.0-beta.2", GitCommit:"0776eab45fe28f02bbeac0f05ae1a203051a21eb", GitTreeState:"clean", BuildDate:"2016-11-24T22:30:23Z", GoVersion:"go1.7.3", Compiler:"gc", Platform:"linux/amd64"}
```

**Environment**:
- **Cloud provider or hardware configuration**: GCE
- **OS** (e.g. from /etc/os-release):
  - master:
    ```
    BUILD_ID=8977.0.0
    NAME="Container-VM Image"
    GOOGLE_CRASH_ID=Lakitu
    VERSION_ID=56
    BUG_REPORT_URL=https://crbug.com/new
    PRETTY_NAME="Google Container-VM Image"
    VERSION=56
    GOOGLE_METRICS_PRODUCT_ID=26
    HOME_URL="https://cloud.google.com/compute/docs/containers/vm-image/"
    ID=gci
    ```
  - minion:
    ```
    PRETTY_NAME="Debian GNU/Linux 7 (wheezy)"
    NAME="Debian GNU/Linux"
    VERSION_ID="7"
    VERSION="7 (wheezy)"
    ID=debian
    ANSI_COLOR="1;31"
    HOME_URL="http://www.debian.org/"
    SUPPORT_URL="http://www.debian.org/support/"
    BUG_REPORT_URL="http://bugs.debian.org/"
    ```
- **Kernel** (e.g. `uname -a`):
  - master: `Linux kubernetes-master 4.4.21+ #1 SMP Thu Nov 10 02:50:15 PST 2016 x86_64 Intel(R) Xeon(R) CPU @ 2.30GHz GenuineIntel GNU/Linux`
  - minion: `Linux kubernetes-minion-group-4klk 3.16.0-4-amd64 #1 SMP Debian 3.16.36-1+deb8u2 (2016-10-19) x86_64 GNU/Linux`
- **Install tools**: `cluster/gce/upgrade.sh`
- **Others**:

**What happened**:

Pods with PVCs get stuck in ContainerCreating after Node gets upgraded. The error is `timeout expired waiting for volumes to attach/mount` as shown below:

```console
$ kubectl describe pod web-1
Name:           web-1
Namespace:      default
Node:           kubernetes-minion-group-54hw/10.128.0.4
Start Time:     Mon, 28 Nov 2016 11:52:11 -0800
Labels:         app=nginx
Status:         Pending
IP:
Controllers:    PetSet/web
Containers:
  nginx:
    Container ID:
    Image:              gcr.io/google_containers/nginx-slim:0.8
    Image ID:   
    Port:               80/TCP
    Requests:
      cpu:              100m
    State:              Waiting
      Reason:           ContainerCreating
    Ready:              False
    Restart Count:      0
    Volume Mounts:
      /usr/share/nginx/html from www (rw)
      /var/run/secrets/kubernetes.io/serviceaccount from default-token-xim26 (ro)
    Environment Variables:      <none>
Conditions:
  Type          Status
  Initialized   True 
  Ready         False 
  PodScheduled  True 
Volumes:
  www:
    Type:       PersistentVolumeClaim (a reference to a PersistentVolumeClaim in the same namespace)
    ClaimName:  www-web-1
    ReadOnly:   false
  default-token-xim26:
    Type:       Secret (a volume populated by a Secret)
    SecretName: default-token-xim26
QoS Class:      Burstable
Tolerations:    <none>
Events:
  FirstSeen     LastSeen        Count   From                                    SubObjectPath       Type             Reason          Message
  ---------     --------        -----   ----                                    -------------       -------- ------          -------
  40m           40m             1       {default-scheduler }                                        Normal           Scheduled       Successfully assigned web-1 to kubernetes-minion-group-54hw
  40m           40m             1       {kubelet kubernetes-minion-group-54hw}  spec.containers{nginx}       Normal          Pulling         pulling image "gcr.io/google_containers/nginx-slim:0.8"
  40m           40m             1       {kubelet kubernetes-minion-group-54hw}  spec.containers{nginx}       Normal          Pulled          Successfully pulled image "gcr.io/google_containers/nginx-slim:0.8"
  40m           40m             1       {kubelet kubernetes-minion-group-54hw}  spec.containers{nginx}       Normal          Created         Created container with docker id d2d15140a74a; Security:[seccomp=unconfined]
  40m           40m             1       {kubelet kubernetes-minion-group-54hw}  spec.containers{nginx}       Normal          Started         Started container with docker id d2d15140a74a
  16m           16m             4       {kubelet kubernetes-minion-group-54hw}                      Warning          FailedMount     MountVolume.SetUp failed for volume "kubernetes.io/secret/270e4f19-b5a4-11e6-8922-42010a800002-default-token-xim26" (spec.Name: "default-token-xim26") pod "270e4f19-b5a4-11e6-8922-42010a800002" (UID: "270e4f19-b5a4-11e6-8922-42010a800002") with: Get https://kubernetes-master/api/v1/namespaces/default/secrets/default-token-xim26: dial tcp: lookup kubernetes-master on 169.254.169.254:53: no such host
  16m           16m             1       {kubelet kubernetes-minion-group-54hw}                      Warning          FailedMount     MountVolume.SetUp failed for volume "kubernetes.io/secret/270e4f19-b5a4-11e6-8922-42010a800002-default-token-xim26" (spec.Name: "default-token-xim26") pod "270e4f19-b5a4-11e6-8922-42010a800002" (UID: "270e4f19-b5a4-11e6-8922-42010a800002") with: Get https://kubernetes-master/api/v1/namespaces/default/secrets/default-token-xim26: dial tcp 10.128.0.2:443: i/o timeout
  15m           15m             3       {kubelet kubernetes-minion-group-54hw}                      Warning          FailedMount     MountVolume.SetUp failed for volume "kubernetes.io/secret/270e4f19-b5a4-11e6-8922-42010a800002-default-token-xim26" (spec.Name: "default-token-xim26") pod "270e4f19-b5a4-11e6-8922-42010a800002" (UID: "270e4f19-b5a4-11e6-8922-42010a800002") with: Get https://kubernetes-master/api/v1/namespaces/default/secrets/default-token-xim26: dial tcp 10.128.0.2:443: getsockopt: connection refused
  9m            9m              1       {kubelet kubernetes-minion-group-54hw}                      Warning          FailedSync      Error syncing pod, skipping: network is not ready: [Kubenet does not have netConfig. This is most likely due to lack of PodCIDR]
  9m            9m              3       {kubelet kubernetes-minion-group-54hw}                      Warning          FailedSync      Error syncing pod, skipping: network is not ready: [Kubenet does not have netConfig. This is most likely due to lack of PodCIDR]
  6m            14s             4       {kubelet kubernetes-minion-group-54hw}                      Warning          FailedMount     Unable to mount volumes for pod "web-1_default(270e4f19-b5a4-11e6-8922-42010a800002)": timeout expired waiting for volumes to attach/mount for pod "web-1"/"default". list of unattached/unmounted volumes=[www]
  6m            14s             4       {kubelet kubernetes-minion-group-54hw}                      Warning          FailedSync      Error syncing pod, skipping: timeout expired waiting for volumes to attach/mount for pod "web-1"/"default". list of unattached/unmounted volumes=[www]
```

**What you expected to happen**:

Upon Node upgrade, Pods should either be terminated (so a controller can recreate them) or they should come back to Running on their own.

**How to reproduce it** (as minimally and precisely as possible):

1.  Start a v1.4 cluster on GCE.

    ```shell
    git checkout release-1.4
    git pull upstream release-1.4
    make clean quick-release
    cluster/kube-up.sh
    ```

1.  Create the following example PetSet and wait for all Pods to be Running.

    ```shell
    kubectl create -f petset.yaml
    ```

    ```yaml
    # A headless service to create DNS records
    apiVersion: v1
    kind: Service
    metadata:
      name: nginx
      labels:
        app: nginx
    spec:
      ports:
      - port: 80
        name: web
      # *.nginx.default.svc.cluster.local
      clusterIP: None
      selector:
        app: nginx
    ---
    apiVersion: apps/v1alpha1
    kind: PetSet
    metadata:
      name: web
    spec:
      serviceName: "nginx"
      replicas: 5
      template:
        metadata:
          labels:
            app: nginx
          annotations:
            pod.alpha.kubernetes.io/initialized: "true"
        spec:
          containers:
          - name: nginx
            image: gcr.io/google_containers/nginx-slim:0.8
            ports:
            - containerPort: 80
              name: web
            volumeMounts:
            - name: www
              mountPath: /usr/share/nginx/html
      volumeClaimTemplates:
      - metadata:
          name: www
          annotations:
            volume.alpha.kubernetes.io/storage-class: anything
        spec:
          accessModes: [ "ReadWriteOnce" ]
          resources:
            requests:
              storage: 1Gi
    ```

1.  Delete PetSet without cascading (keep the Pods).

    ```shell
    kubectl delete petset web --cascade=false
    ```

1.  Upgrade master and nodes to `v1.5.0-beta.2` with `cluster/gce/upgrade.sh`.

    ```shell
    git checkout release-1.5
    git pull upstream release-1.5
    cluster/gce/upgrade.sh v1.5.0-beta.2
    ```

1.  Check for Pods stuck in ContainerCreating.

    ```
    NAME      READY     STATUS              RESTARTS   AGE
    web-0     0/1       ContainerCreating   0          42m
    web-4     0/1       ContainerCreating   0          42m
    ```

**Anything else do we need to know**:

After the above steps, some Pods may end up terminating. It seems to be random which ones terminate and which ones get stuck in ContainerCreating.

Manually deleting and recreating all the stuck Pods after the Nodes are upgraded allows them to correctly attach to the preexisting PVCs and become Running.

**Comment by enisoc** [comment]
@janetkuo @saad-ali Could one of you apply the labels we discussed? I don't have access to do it.

**Comment by janetkuo** [comment]
@kubernetes/sig-storage @jingxu97 PTAL on this, it's 1.5 release blocker

**Comment by saad-ali** [comment]
Possibly the same root cause as https://github.com/kubernetes/kubernetes/issues/37586

**Comment by dims** [comment]
@saad-ali : need an assignee

**Comment by saad-ali** [comment]
@rkouj is investigating/working on a fix in https://github.com/kubernetes/kubernetes/issues/37586

**Comment by dims** [comment]
Thanks @saad-ali @rkouj

**Comment by rkouj** [comment]
`WIP`
https://github.com/kubernetes/kubernetes/pull/37727

**Comment by dims** [comment]
@rkouj - reasonably sure that we unblock this issue with #37727 right?

**Comment by jingxu97** [comment]
@rkouj yes, we think so.

**Comment by enisoc** [comment]
This seems to be fixed on master. I retried the upgrade, this time from `v1.4.6` to `v1.6.0-alpha.0.1228+2212c421f6e10e` (ci/latest). All pods that didn't get deleted progressed through ContainerCreating into Running this time.

There was one pod that got deleted and then after my controller recreated it on a different node, it got stuck in ContainerCreating with the messsage:

```
Failed to attach volume "pvc-e8d989f7-b8b4-11e6-81ec-42010a800002" on node "kubernetes-minion-group-ctw8" with: googleapi: Error 400: The disk resource 'kubernetes-dynamic-pvc-e8d989f7-b8b4-11e6-81ec-42010a800002' is already being used by 'kubernetes-minion-group-z72i'
```

I checked Cloud Console, and indeed the disk was marked as being in use by the previous node. The situation resolved itself after 8min and then everything became healthy. I'm not sure if it's expected to take that long, but overall my use case now works. I'll try again when the fix makes it into the release-1.5 branch.

**Comment by jingxu97** [comment]
Hi @enisoc thanks a lot for trying out the fix. About the 8 min delay, could you please share your kube-controller-manager log so that we can check it? Thanks!

**Comment by rkouj** [comment]
The cherry pick PR is out and should be merged soon ~O(hours).
https://github.com/kubernetes/kubernetes/pull/37973

@enisoc Would it make sense to create a separate issue for the 8 min delay ? This would help keeping things separate and I can close this. Also, is that issue a release blocker ?

**Comment by enisoc** [comment]
@jingxu97 @rkouj Unfortunately I had torn down the cluster and didn't save the logs. I will try again and create a separate issue if I encounter the delay due to "already being used" again. It's not a release blocker for my use case (checking that PetSet->StatefulSet upgrades work) because it resolved itself, but others may consider the delay important. We can discuss on the separate issue if I'm able to reproduce it.

**Comment by rkouj** [comment]
Thanks for letting us know!

**Comment by enisoc** [comment]
I verified my repro case is also fixed for the release-1.5 branch following the cherrypick, by using `v1.5.0-beta.2.54+92991e70b8dd5e` as the upgrade target.

**Comment by dims** [comment]
https://github.com/kubernetes/kubernetes/pull/37727 and https://github.com/kubernetes/kubernetes/pull/37973 are merged. So let's close this issue now. thanks to @enisoc for verifying as well!
